runtime/vm/stringhelpers.cpp

```diff
@@ -258,179 +258,96 @@ compareStringToUTF8(J9VMThread *vmThread, j9object_t string, UDATA translateDots
- * Copy a Unicode String to a UTF8 data buffer without NULL termination.
- * dest is assumed to have enough length - the easiest way to ensure this is to pass a buffer with 1.5 * string length in it
- *
- * @param[in] vmThread the current J9VMThread
- * @param[in] string a string object to be copied, it can't be NULL
- * @param[in] dest a utf8 data buffer
- *
- * @return UDATA_MAX if a failure occurred, otherwise the number of utf8 data copied excluding null termination
- */
-UDATA
-copyFromStringIntoUTF8(J9VMThread * vmThread, j9object_t string, char * dest)
-{
-	return copyStringToUTF8Helper(vmThread, string, FALSE, J9_STR_NONE, (U_8*)dest, UDATA_MAX);
+	return returnLength;
 }
 
-
-/**
- * Copy a string object to a UTF8 data buffer, optionally to prepend a string before it
- *
- * ***The caller must free the memory from this pointer if the return value is NOT the buffer argument ***
- *
- * @param[in] vmThread the current J9VMThread
- * @param[in] string a string object to be copied, it can't be NULL
- * @param[in] stringFlags the flag to determine performing '.' --> '/'
- * @param[in] prependStr the string to be prepended before the string object to be copied
- * 				it can't be NULL but can be an empty string ""
- * @param[in] buffer the buffer for the string
- * @param[in] bufferLength the buffer length, not expected to larger than 64K
- *
- * @return a char pointer to the string (with NULL termination)
- */
 char*
-copyStringToUTF8WithMemAlloc(J9VMThread *vmThread, j9object_t string, UDATA stringFlags, const char *prependStr, char *buffer, UDATA bufferLength)
+copyStringToUTF8WithMemAlloc(J9VMThread *vmThread, j9object_t string, UDATA stringFlags, const char *prependStr, UDATA prependStrLength, char *buffer, UDATA bufferLength, UDATA *utf8Length)
 {
 	Assert_VM_notNull(prependStr);
 	Assert_VM_notNull(string);
 
 	char *strUTF = NULL;
-	UDATA stringLen = getStringUTF8Length(vmThread, string);
-	UDATA prependStrLen = strlen(prependStr);
-	UDATA length = stringLen + prependStrLen + 1;
+	UDATA stringLen = J9VMJAVALANGSTRING_LENGTH(vmThread, string) * 3;
+	UDATA length = stringLen + prependStrLength;
+
+	if ((stringFlags & J9_STR_NULL_TERMINATE_RESULT) != 0) {
+		++length;
+	}
 
 	PORT_ACCESS_FROM_VMC(vmThread);
 
 	if (length > bufferLength) {
 		strUTF = (char *)j9mem_allocate_memory(length, OMRMEM_CATEGORY_VM);
 	} else {
 		strUTF = buffer;
-		length = bufferLength;
 	}
 	if (NULL != strUTF) {
-		if (0 < prependStrLen) {
-			memcpy(strUTF, prependStr, prependStrLen);
+		UDATA computedUtf8Length = 0;
+
+		if (0 < prependStrLength) {
+			memcpy(strUTF, prependStr, prependStrLength);
 		}
-		if (UDATA_MAX == copyStringToUTF8Helper(
-			vmThread, string, TRUE, stringFlags, (U_8 *)(strUTF + prependStrLen), length - prependStrLen)
-		) {
-			if (buffer != strUTF) {
-				j9mem_free_memory(strUTF);
-			}
-			return NULL;
+
+		computedUtf8Length = copyStringToUTF8Helper(vmThread, string, stringFlags, (U_8 *)(strUTF + prependStrLength), length);
```

Comment:
Instead of length this should pass stringLen or bufferLength - prependStrLength, whichever is smaller.

Reply:
Fixed in a544cba. Good catch!